sw/device/tests/sim_dv/pwrmgr_normal_sleep_all_wake_ups.c

```diff
@@ -0,0 +1,344 @@
+  };
+  CHECK_DIF_OK(dif_pinmux_wakeup_detector_enable(dif, kPinmuxWkupDetector5,
+                                                 detector_cfg));
+}
+
+/**
+ * usb config for test #4
+ * . Fake low power entry through usb
+ * . Force usb to output suspend indication
+ * (*dif) handle is not used but leave as is
+ * to be called from execute_test
+ */
+static void prgm_usb_wakeup(void *dif) {
+  usbdev_set_wake_module_active(true);
+  usbdev_force_dx_pullup(kDpSel, true);
+  usbdev_force_dx_pullup(kDnSel, false);
+
+  // give the hardware a chance to recognize the wakeup values are the same
+  busy_spin_micros(20);  // 20us
+}
+
+/**
+ * aon timer config for test #5
+ * set wakeup signal in 50us
+ */
+static void prgm_aontimer_wakeup(void *dif) {
+  aon_timer_testutils_wakeup_config(dif, 10);
+}
+
+/**
+ * sensor ctrl config for test #6
+ * setup event trigger0
+ */
+static void prgm_sensor_ctrl_wakeup(void *dif) {
+  CHECK_DIF_OK(
+      dif_sensor_ctrl_set_ast_event_trigger(dif, 0, kDifToggleEnabled));
+}
+static const test_wakeup_sources_t kTestWakeupSources[] = {
+    {
+        .name = "SYSRST_CTRL",
+        .dif_handle = &sysrst_ctrl,
+        .wakeup_src = kDifPwrmgrWakeupRequestSourceOne,
+        .config = prgm_sysrst_ctrl_wakeup,
+    },
+    {
+        .name = "ADC_CTRL",
+        .dif_handle = &adc_ctrl,
+        .wakeup_src = kDifPwrmgrWakeupRequestSourceTwo,
+        .config = prgm_adc_ctrl_wakeup,
+    },
+    {
+        .name = "PINMUX",
+        .dif_handle = &pinmux,
+        .wakeup_src = kDifPwrmgrWakeupRequestSourceThree,
+        .config = prgm_pinmux_wakeup,
+    },
+    {
+        .name = "USB",
+        .dif_handle = &usbdev,
+        .wakeup_src = kDifPwrmgrWakeupRequestSourceFour,
+        .config = prgm_usb_wakeup,
+    },
+    {
+        .name = "AONTIMER",
+        .dif_handle = &aon_timer,
+        .wakeup_src = kDifPwrmgrWakeupRequestSourceFive,
+        .config = prgm_aontimer_wakeup,
+    },
+    {
+        .name = "SENSOR_CTRL",
+        .dif_handle = &sensor_ctrl,
+        .wakeup_src = kDifPwrmgrWakeupRequestSourceSix,
+        .config = prgm_sensor_ctrl_wakeup,
+    },
+};
+
+/**
+ * External interrupt handler.
+ */
+void ottf_external_isr(void) {
+  dif_pwrmgr_irq_t irq_id;
+  top_earlgrey_plic_peripheral_t peripheral;
+
+  isr_testutils_pwrmgr_isr(plic_ctx, pwrmgr_isr_ctx, &peripheral, &irq_id);
+
+  // Check that both the peripheral and the irq id is correct
+  CHECK(peripheral == kTopEarlgreyPlicPeripheralPwrmgrAon,
+        "IRQ peripheral: %d is incorrect", peripheral);
+  CHECK(irq_id == kDifPwrmgrIrqWakeup, "IRQ ID: %d is incorrect", irq_id);
+}
+
+static void execute_test(uint32_t wakeup_source) {
+  // Configure wakeup device
+  kTestWakeupSources[wakeup_source].config(
+      kTestWakeupSources[wakeup_source].dif_handle);
+  // Normal sleep
+  pwrmgr_testutils_enable_low_power(
+      &pwrmgr, kTestWakeupSources[wakeup_source].wakeup_src,
+      kDifPwrmgrDomainOptionMainPowerInLowPower);
```

Comment:
two cent idea : how about considering low power for each sub-domain such as USB? For example, we can configure kDifPwrmgrDomainOptionUsbClockInLowPower | kDifPwrmgrDomainOptionMainPowerInLowPower for usb wake-up.

Reply:
you are suggesting randomly enable any of {CORE_CLK_EN, IO_CLK_EN, USB_CLK_EN_LP} during normal sleep state, correct? I can add that.
@tjaychen WDYT?

Reply:
It's not a random enable. When each wake-up src has its corresponding domain like USB, it would be better to make the domain sleep, too.

Reply:
Just talked with @jdonjdon. All clks are off and on together except the MainPower domain. Thanks @jdonjdon

Reply:
Sorry, this is not clear. In shallow sleep we can program any domain to be off per pwrmgr CONTROL csr, yet the wakeup should trigger anyway.

Reply:
Could you clarify what is meant in @jeoongp comment from June 6?

Reply:
Sure, below is brief discussion I had with @jeoongp .

From PWRMGR.CONTROL (https://docs.opentitan.org/hw/ip/pwrmgr/doc/index.html#register-table),
we can selectively shut down clock among 4 groups (main, core, io, and usb).
For normal sleep, main (bit 8) should be on but the others 'can be' on/off.
When clocks are off, they can be back with wake-up EXCEPT usb.
USB needs manual csr program to turn on.

Reply:
Yet in line 252 you are calling pwrmgr_testutils_enable_low_power with domain_config set to pwrmgr_testutils_enable_low_power, meaning all clocks are disabled in normal sleep. I suppose this is more likely to fail than having some clocks on.

I don't understand the comment about USB: it runs in the active state depending on kDifPwrmgrDomainOptionUsbClockInActivePower, so it all depends on the args to pwrmgr_testutils_enable_low_power.

Reply:
updated as suggested.